crates/ruff_linter/src/rules/flake8_pyi/rules/generic_not_last_base_class.rs

```diff
@@ -0,0 +1,130 @@
+/// in stubs it is purely enforced for stylistic consistency.
+///
+/// For example:
+/// ```python
+/// class LinkedList(Generic[T], Sized):
+///     def push(self, item: T) -> None:
+///         self._items.append(item)
+///
+/// class MyMapping(
+///     Generic[K, V],
+///     Iterable[Tuple[K, V]],
+///     Container[Tuple[K, V]],
+/// ):
+///     ...
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class LinkedList(Sized, Generic[T]):
+///     def push(self, item: T) -> None:
+///         self._items.append(item)
+///
+/// class MyMapping(
+///     Iterable[Tuple[K, V]],
+///     Container[Tuple[K, V]],
+///     Generic[K, V],
+/// ):
+///     ...
+/// ```
+/// ## References
+/// - [`typing.Generic` documentation](https://docs.python.org/3/library/typing.html#typing.Generic)
+///
+/// [1]: https://github.com/python/cpython/issues/106102
+#[violation]
+pub struct GenericNotLastBaseClass;
+
+impl Violation for GenericNotLastBaseClass {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`Generic[]` should always be the last base class")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Move `Generic[]` to the end".to_string())
+    }
+}
+
+/// PYI059
+pub(crate) fn generic_not_last_base_class(checker: &mut Checker, class_def: &StmtClassDef) {
+    let Some(bases) = class_def.arguments.as_deref() else {
+        return;
+    };
+
+    let semantic = checker.semantic();
+    if !semantic.seen_typing() {
+        return;
+    }
+
+    let Some(last_base) = bases.args.last() else {
+        return;
+    };
+
+    let mut generic_base_iter = bases
+        .args
+        .iter()
+        .filter(|base| semantic.match_typing_expr(map_subscript(base), "Generic"));
+
+    let Some(generic_base) = generic_base_iter.next() else {
+        return;
+    };
+
+    // If Generic[] exists, but is the last base, don't raise issue.
+    if generic_base.range() == last_base.range() {
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(GenericNotLastBaseClass, bases.range());
+
+    // No fix if multiple generics are seen in the class bases.
+    if generic_base_iter.next().is_none() {
+        diagnostic.set_fix(generate_fix(generic_base, last_base, checker.locator()));
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+fn generate_fix(generic_base: &Expr, last_base: &Expr, locator: &Locator) -> Fix {
+    let comma_after_generic_base = generic_base.end().to_usize()
+        + locator
+            .after(generic_base.end())
+            .find(',')
+            .expect("Comma must always exist after generic base");
+
+    let last_whitespace = (comma_after_generic_base + 1)
+        + locator.contents()[comma_after_generic_base + 1..]
+            .bytes()
+            .position(|b| !b.is_ascii_whitespace())
+            .expect("Non whitespace character must always exist after Generic[]");
```

Comment:
I don't really understand what's going on here :(

Would you mind adding some comments to this function to explain how it works (preferably with some examples)?

Reply:
I'm trying to find the first index in the source code after `Generic[...],`, which doesn't contain a whitespace.

Basically, deleting just `Generic[T]` and the comma after it would leave a trailing whitespace around (usually a space).

Without the whitespace detection, this code:

```python
class Foo(Generic[T], Bar):
```

would be autofixed into:

```python
class Foo( Bar, Generic[T]):
```

i.e. The space before `Bar` was not deleted.

Reply:
I'm honestly not sure if this is the best way to write this, please feel free to refactor or comment on this.

Reply:
I think deleting a positional argument from a function call during autofix is common enough that once we finalize on the way it should be done this can be lifted to make a helper instead.

Reply:
Ah... in fact... https://github.com/astral-sh/ruff/blob/12b5c3a54c0a34bae9dfcfeaf27961f4432d5623/crates/ruff_linter/src/fix/edits.rs#L155-L237

Reply:
I'm so sorry, I completely forgot that we had those helpers!

Reply:
I'll go ahead and refactor it.